### PR TITLE
Update ICRC-28 and 34

### PR DESCRIPTION
**Pull request by dostro**
Updated with comments from prior PR and today's WG session

### CODE REVIEW

File: topics/icrc_34_get_delegation.md [hunk 1]
```diff
@@ -1,50 +1,54 @@
...
+[ICRC-28](./icrc_28_trusted_origins.md) standard.
 
-**Name and Scope:** `icrc34_get_global_delegation`
+Relying Party Delegations **MUST** be restricted by the signer to individual relying parties.
 
-**Prerequisite:** Active session with granted permission scope `icrc34_get_global_delegation` or `*`.
+Signers **SHOULD** give users the choice to authenticate with their wallets (if available) or with an app-specific identifier.
 
-## Scope (according to the [ICRC-25 standard](./icrc_25_signer_interaction_standard.md))
+Relying parties **SHOULD** remove targets if they want to isolate users from the rest of the ecosystem.
```
Comment (sea-snake) [inline topics/icrc_34_get_delegation.md:39]: This could be one purpose, but shouldn't be the main purpose of Relying Party delegations. We obviously don't want to isolate users intentionally.

Also a dapp that uses Relying Party delegations could still request the list of accounts and communicate the principal from this list as user identifier across it's platform, making the user data thus non isolated from the rest of the ecosystem.

Also using a Relying Party delegation to isolate users would run into issues, since it's not an Account delegation for a principal returned in the list of accounts, it can't hold assets that can be managed by the wallet.

Instead I would recommend to use Relying Party delegations for more technical use cases, which I tried to suggest above in the summary. 

Nonetheless, above summary should attempt to clarify the difference between a Account and Relying Party delegation. This part of the spec should focus instead on clearly indicating how to get one type of delegation or the other. 

```suggestion
Relying parties **MUST** remove targets if they want to receive a Relying Party delegation.
```

File: topics/icrc_34_get_delegation.md [hunk 2]
```diff
@@ -1,50 +1,54 @@
...
-## Method
+If a relying party wants to receive an Account Delegation, the `icrc34_delegation` request **MUST** include canisters it
+controls as `targets` where each canister **MUST** implement the `icrc28_trusted_origins` endpoint as per the 
+[ICRC-28](./icrc_28_trusted_origins.md) standard.
 
-**Name and Scope:** `icrc34_get_global_delegation`
+Relying Party Delegations **MUST** be restricted by the signer to individual relying parties.
 
-**Prerequisite:** Active session with granted permission scope `icrc34_get_global_delegation` or `*`.
+Signers **SHOULD** give users the choice to authenticate with their wallets (if available) or with an app-specific identifier.
```
Comment (sea-snake) [inline topics/icrc_34_get_delegation.md:37]: I would personally consider keeping this out of the spec, a signer could decide to not give the user choice but decide based on the type of delegation requested. Which in most signers seems a more straightforward implementation.

Basically giving a user the choice to return a RP delegation when an Account delegation is requested is an advanced implementation to allow for anonymity but is also has technical challenges like what to return in the accounts request etc.

I can understand the reasons for adding it to the spec, but I would at least change it to `COULD` since the main goal here is probably to indicate that signers _could_ return a RP delegation when a Account delegation was requested.

```suggestion
Signers **COULD** give users the choice to authenticate with their relying party specific identifier instead of accounts (if available).
```

File: topics/icrc_34_get_delegation.md [hunk 3]
```diff
@@ -1,50 +1,54 @@
...
-trust the relying party following the [ICRC-28](./icrc_28_trusted_origins.md) standard.
+When a relying party wants to remove wallet approval prompts and sign on behalf of its users, it uses the icrc34_delegation 
+method described below. This method generates a delegation chain, an object with a stable identifier, allowing the relying 
+party to sign transactions on behalf of the user.
 
-- Due to the canister target restriction, only the canisters that the relying party owns can be called on behalf of the user.
-- A new delegation chain must be obtained for every new canister created by the relying party.
-- The identity is the same across all relying parties, the user is not anonymous between different relying parties.
+The delegation chain's stable identifier can take two forms: an account that may include ICP, ICRC-1, and other assets 
+(Account Delegation), or an account designed exclusively for the relying party (Relying Party Delegation).
```
Comment (sea-snake) [inline topics/icrc_34_get_delegation.md:29]: Since ICP is an ICRC-1 token, I would suggest removing ICP from the example.

Also would suggest below change with a numbered list to clearly list each type one by one.

```suggestion
The delegation chain's stable identifier can be either: 
1. **Account Delegation**: an identity that may hold tokens and other assets.
2. **Relying Party Delegation**: an identity designed exclusively for the relying party (Relying Party Delegation).
```

File: topics/icrc_34_get_delegation.md [hunk 4]
```diff
@@ -1,50 +1,54 @@
...
 ## Summary
 
-When a relying party wants to authenticate as a user, it uses
-a [session key](https://internetcomputer.org/docs/current/references/ic-interface-spec/#signatures), and
-below `icrc34_get_global_delegation` method to obtain a delegation chain that allows the session key to sign for the
-user's global identity. The obtained delegation chain **MUST** be restricted by the signer to canister targets that
-trust the relying party following the [ICRC-28](./icrc_28_trusted_origins.md) standard.
+When a relying party wants to remove wallet approval prompts and sign on behalf of its users, it uses the icrc34_delegation 
+method described below. This method generates a delegation chain, an object with a stable identifier, allowing the relying 
+party to sign transactions on behalf of the user.
```
Comment (sea-snake) [inline topics/icrc_34_get_delegation.md:26]: I would rephrase this as `avoid going through the signer and user approval for each individual call`.
Also the fact that the relying party does not need to stay connected to the signer when it has a delegation is an interesting aspect that might be worth mentioning.

```suggestion
When a relying party wants to avoid going through the signer and user approval for each individual call, it can use the icrc34_delegation method described below. 

This method generates a delegation chain, an object with a stable identifier, allowing the relying party to sign transactions on behalf of the user without having to go through the signer and user approval. This also means the signer does not need to stay connected to the signer to make calls on behalf of the user.
```

File: topics/icrc_34_get_delegation.md [hunk 5]
```diff
@@ -1,50 +1,54 @@
...
 
-- Due to the canister target restriction, only the canisters that the relying party owns can be called on behalf of the user.
-- A new delegation chain must be obtained for every new canister created by the relying party.
-- The identity is the same across all relying parties, the user is not anonymous between different relying parties.
+The delegation chain's stable identifier can take two forms: an account that may include ICP, ICRC-1, and other assets 
+(Account Delegation), or an account designed exclusively for the relying party (Relying Party Delegation).
```
Comment (sea-snake) [inline topics/icrc_34_get_delegation.md:30]: The use case of a Relying Party Delegation is still unclear, maybe below explanation might help?

```suggestion

Example use cases of a Relying Party Delegation:
- Signer could give the user a choice to stay anonymous with a service.
- Differentiate between calls made with user approval (Account identifier) and without user approval (Relying Party Delegation identifier). Allowing for fine grained security levels per identifier.
- Exclusive identifier within the Relying Party platform to stay isolated from identifiers of other Relying Party platforms.

```

**Review by sea-snake — COMMENTED** [review]
Some feedback and suggested changes.

File: topics/icrc_34_get_delegation.md (hunk 1 above)
Comment (dostro) [reply topics/icrc_34_get_delegation.md:39]: Good point.

Only thought is the difference between **must** and **should** in this context - if a relying party wants to receive an RP delegation, and they **could** receive one even if targets are included, does the clause here need to be "must"?

Comment (sea-snake) [reply topics/icrc_34_get_delegation.md:45]: I would say yes it should be `MUST` since a RP delegation is not guaranteed when targets are defined but a side effect of the user deciding to be anonymous or targets not passing ICRC-28 validation.

So to be _guaranteed_ to receive a RP delegation, a RP must remove the targets.

Comment (sea-snake) [reply topics/icrc_34_get_delegation.md:45]: Added guaranteed to the sentence since that should probably clarify it.

File: topics/icrc_28_trusted_origins.md [hunk 6]
```diff
@@ -115,57 +79,4 @@ sequenceDiagram
     * If the origin is trusted by all targets, continue with step 4.
     * If the origin is not trusted by all targets, the signer returns an error to the relying party. No further steps
       are executed.
```
Comment (sea-snake) [inline topics/icrc_28_trusted_origins.md:81]:
```suggestion
    * If the origin is trusted by all targets, continue with step 4a.
    * If the origin is not trusted by all targets, continue with step 4b.
```

File: topics/icrc_28_trusted_origins.md [hunk 7]
```diff
@@ -115,57 +79,4 @@ sequenceDiagram
...
-    3. The relying party origin must be within all decoded trusted origins responses.
-5. The delegation request is signed and transferred to the chain connected component.
-    * The expiry of the delegation request is set to the most recent response certificate `time` within the trusted
-      origins responses plus a reasonable session length that is less than or equal to the maxTimeToLive value in the
-      delegation request.
-6. The signed delegation is returned to the relying party.
-
-> It's recommended to have the Chain Connected Signer Component verify the delegation request beforehand as described in
-> above _hot_ signer interaction use-case. To reject invalid requests early before interacting with the cold signer.
+4. The signed wallet delegation is returned to the relying party.
```
Comment (sea-snake) [inline topics/icrc_28_trusted_origins.md:82]:
```suggestion
4a. The signed account delegation is returned to the relying party.
4b. The signed relying party delegation is returned to the relying party.
```

**Review by sea-snake — COMMENTED** [review]
Last few details I noticed in ICRC-28.

**Review by sea-snake — APPROVED** [review]
PR seems good to go my end 🎉

File: topics/icrc_34_delegation.md [hunk 8]
```diff
@@ -1,50 +1,60 @@
...
-below `icrc34_get_global_delegation` method to obtain a delegation chain that allows the session key to sign for the
-user's global identity. The obtained delegation chain **MUST** be restricted by the signer to canister targets that
-trust the relying party following the [ICRC-28](./icrc_28_trusted_origins.md) standard.
+When a relying party wants to avoid going through the signer and user approval for each individual call, it can use the icrc34_delegation method described below. 
 
-- Due to the canister target restriction, only the canisters that the relying party owns can be called on behalf of the user.
-- A new delegation chain must be obtained for every new canister created by the relying party.
-- The identity is the same across all relying parties, the user is not anonymous between different relying parties.
+This method generates a delegation chain, an object with a stable identifier, allowing the relying party to sign transactions on behalf of the user without having to go through the signer and user approval. This also means the signer does not need to stay connected to the signer to make calls on behalf of the user.
```
Comment (frederikrothenberger) [inline topics/icrc_34_delegation.md:26]: Cutting out some repetition:

```suggestion
This method generates a delegation chain, an object with a stable identifier, allowing the relying party to independently sign transactions on behalf of the user. This also means the signer does not need to stay connected to the signer to make calls on behalf of the user.
```

File: topics/icrc_34_delegation.md [hunk 9]
```diff
@@ -1,50 +1,60 @@
...
 
-## Method
+The delegation chain's stable identifier can be either: 
+1. **Account Delegation**: an identity that may hold tokens and other assets.
+2. **Relying Party Delegation**: an identity designed exclusively for the relying party (Relying Party Delegation).
```
Comment (frederikrothenberger) [inline topics/icrc_34_delegation.md:30]: I think "designed" is the wrong term? 

```suggestion
2. **Relying Party Delegation**: an identity created exclusively for the relying party (Relying Party Delegation).
```

File: topics/icrc_34_delegation.md [hunk 10]
```diff
@@ -1,50 +1,60 @@
...
+When a relying party wants to avoid going through the signer and user approval for each individual call, it can use the icrc34_delegation method described below. 
 
-- Due to the canister target restriction, only the canisters that the relying party owns can be called on behalf of the user.
-- A new delegation chain must be obtained for every new canister created by the relying party.
-- The identity is the same across all relying parties, the user is not anonymous between different relying parties.
+This method generates a delegation chain, an object with a stable identifier, allowing the relying party to sign transactions on behalf of the user without having to go through the signer and user approval. This also means the signer does not need to stay connected to the signer to make calls on behalf of the user.
 
-## Method
+The delegation chain's stable identifier can be either: 
+1. **Account Delegation**: an identity that may hold tokens and other assets.
```
Comment (frederikrothenberger) [inline topics/icrc_34_delegation.md:29]: That property is not exclusive to account delegations, see the current ecosystem of in-dapp wallets.

I think the crucial point is that it is stable across many relying parties (but restricted by `targets`).

And further, if it does hold assets, the delegation _must not_ give the relying party access to them.

Maybe: "Restricted access to the signers identity, such that it is stable across many relying parties but cannot be used to operate on tradable assets and shared infrastructure"

File: topics/icrc_34_delegation.md [hunk 11]
```diff
@@ -1,50 +1,60 @@
...
+2. **Relying Party Delegation**: an identity designed exclusively for the relying party (Relying Party Delegation).
 
-**Name and Scope:** `icrc34_get_global_delegation`
+Example use cases of a Relying Party Delegation:
+- Signer could give the user a choice to stay anonymous with a service.
+- Differentiate between calls made with user approval (Account identifier) and without user approval (Relying Party Delegation identifier). Allowing for fine grained security levels per identifier on the Relying Party platform.
```
Comment (frederikrothenberger) [inline topics/icrc_34_delegation.md:34]: This is true for ledgers as well (i.e. `icrc2_approve`):

```suggestion
- Differentiate between calls made with user approval (Account identifier) and without user approval (Relying Party Delegation identifier), allowing for fine grained security levels per identifier.
```

File: topics/icrc_34_delegation.md [hunk 12]
```diff
@@ -1,50 +1,60 @@
...
+- Exclusive identifier within the Relying Party platform to stay isolated from identifiers of other Relying Party platforms.
 
-**Prerequisite:** Active session with granted permission scope `icrc34_get_global_delegation` or `*`.
+If a relying party wants to receive an Account Delegation, the `icrc34_delegation` request **MUST** include canisters it
```
Comment (frederikrothenberger) [inline topics/icrc_34_delegation.md:37]: The "it controls" part is not a requirement per ICRC-28, from the ICRC-28 perspective the developer must simply make sure it is not malicious.
If you want to explicitly restrict ICRC-28 to only be used by canisters under the exact same control structure, then that should be reflected there as well.

File: topics/icrc_34_delegation.md [hunk 13]
```diff
@@ -1,50 +1,60 @@
...
+controls as `targets` where each canister **MUST** implement the `icrc28_trusted_origins` endpoint as per the 
+[ICRC-28](./icrc_28_trusted_origins.md) standard.
 
-## Scope (according to the [ICRC-25 standard](./icrc_25_signer_interaction_standard.md))
+Relying Party Delegations **MUST** be restricted by the signer to individual relying parties.
```
Comment (frederikrothenberger) [inline topics/icrc_34_delegation.md:41]:
```suggestion
Relying Party Delegations **MUST** be for identities exclusive to each individual relying party. This property **MUST** be enforced by the signer.
```

File: topics/icrc_34_delegation.md [hunk 14]
```diff
@@ -1,50 +1,60 @@
...
+
+Signers **COULD** give users the choice to authenticate with their relying party specific identifier instead of accounts (if available).
```
Comment (frederikrothenberger) [inline topics/icrc_34_delegation.md:43]:
```suggestion
Signers **MAY** give users the choice to authenticate with their relying party specific identifier instead of accounts (if available).
```

File: topics/icrc_34_delegation.md [hunk 15]
```diff
@@ -1,50 +1,60 @@
...
+
+Relying parties **MUST** remove targets if they want to be guaranteed to receive a Relying Party delegation.
```
Comment (frederikrothenberger) [inline topics/icrc_34_delegation.md:45]: As this is about conveying a preference I don't think the uppercase MUST is warranted here.

```suggestion
Relying parties must not include `targets` in the request if they want to be guaranteed to receive a Relying Party delegation.
```

File: topics/icrc_34_delegation.md [hunk 16]
```diff
@@ -1,50 +1,60 @@
...
+
+## Method
+
+**Name:** `icrc34_delegation`
 
-**Scope:** `icrc34_get_global_delegation`
+**Prerequisite:** Active session with granted permission scope `icrc34_delegation`.
```
Comment (frederikrothenberger) [inline topics/icrc_34_delegation.md:51]: I realize I need to make this change in the other extensions too...

```suggestion
**Prerequisite:** Granted permission scope `icrc34_delegation`.
```

File: topics/icrc_34_delegation.md [hunk 17]
```diff
@@ -54,34 +64,38 @@ trust the relying party following the [ICRC-28](./icrc_28_trusted_origins.md) st
...
+- `targets` (`text` `array` optional): A list of target canister ids (textual representation) Account Delegations can execute transactions
+for on behalf of the user.
+- `maxTimeToLive` (`text` optional): Expiration of the delegation in nanoseconds, though signer can still choose to
 return a delegation with a shorter expiration.
 
-### Example RPC Request
+### Example RPC Delegation Request
+
+Requests for a delegation can optionally include `targets`, enabling signers to verify the security of returning an Account 
+Delegation (via ICRC-28). If the array of `targets` is empty, signers **SHOULD** only offer users the option to connect
```
Comment (frederikrothenberger) [inline topics/icrc_34_delegation.md:90]: My intention was to make it clearer that this is _not_ the case where targets is `undefined`.

```suggestion
Delegation (via ICRC-28). If the array of `targets` is set to an empty array, signers **SHOULD** only offer users the option to connect 
```

File: topics/icrc_34_delegation.md [hunk 18]
```diff
@@ -136,38 +151,16 @@ the [IC interface specification, authentication section](https://internetcompute
...
 
-1. The relying party sends a `icrc34_get_global_delegation` request to the signer.
+1. The relying party sends a `icrc34_delegation` request to the signer.
 2. Upon receiving the request, the signer validates whether it can process the message.
     * If the relying party has not been granted the permission to request the action, the signer sends a response with
       an error back to the relying party.
     * The relying party must make sure that the request complies with scope targets restriction.
-3. The signer tries to retrieve and verify the trusted origins according to the [ICRC-28](icrc_28_trusted_origins.md)
-   specification.
+3. If the request includes targets, the signer **MAY** try to retrieve and verify the trusted origins according to the [ICRC-28](icrc_28_trusted_origins.md) specification.
```
Comment (frederikrothenberger) [inline topics/icrc_34_delegation.md:159]:
```suggestion
3. If the request includes targets, the signer **MAY** offer issuing an account delegation. If it does, it **MUST** retrieve and verify the trusted origins according to the [ICRC-28](icrc_28_trusted_origins.md) specification.
```

File: topics/icrc_34_delegation.md [hunk 19]
```diff
@@ -136,38 +151,16 @@ the [IC interface specification, authentication section](https://internetcompute
...
-        alt Origin is trusted by all targets canisters
-            S ->> RP: Signed delegation
-        else
-            S ->> RP: Error response
-        end
-    end
-```
+      not within any of the trusted origin lists, the signer does not give users the ability to continue with the Account Delegation.
+4. The signer **MAY** display all the available delegations the user can continue with, in which case a user would select one.
+5. The signer returns a signed delegation to the relying party.
```
Comment (frederikrothenberger) [inline topics/icrc_34_delegation.md:163]:
```suggestion
5. The signer returns the signed delegation to the relying party.
```

File: topics/icrc_34_delegation.md [hunk 20]
```diff
@@ -136,38 +151,16 @@ the [IC interface specification, authentication section](https://internetcompute
...
-3. The signer tries to retrieve and verify the trusted origins according to the [ICRC-28](icrc_28_trusted_origins.md)
-   specification.
+3. If the request includes targets, the signer **MAY** try to retrieve and verify the trusted origins according to the [ICRC-28](icrc_28_trusted_origins.md) specification.
     * If the trusted origins cannot be retrieved for any of the given delegations targets or the relying party origin is
-      not within any of the trusted origin lists, the signer rejects the request and send a response with an error code
-      back to the relying party.
-4. The signer returns a signed delegation to the relying party
-
-```mermaid
-sequenceDiagram
```
Comment (frederikrothenberger) [inline topics/icrc_34_delegation.md:152]: Why did you drop the sequence diagram? I think it should still be included here.

File: topics/icrc_34_delegation.md [hunk 21]
```diff
@@ -1,50 +1,60 @@
...
   * [Summary](#summary)
   * [Method](#method)
   * [Scope (according to the ICRC-25 standard)](#scope-according-to-the-icrc-25-standard)
-    * [Example RPC Request Permission](#example-rpc-request-permission)
-  * [Request](#request)
-    * [Example RPC Request](#example-rpc-request)
-    * [Response](#response)
-    * [Example RPC Response](#example-rpc-response)
+    * [Example RPC Permission Request](#example-rpc-permission-request)
+  * [icrc-25-supported-standards](#icrc-25-supported-standards)
```
Comment (frederikrothenberger) [inline topics/icrc_34_delegation.md:13]:
```suggestion
  * [icrc25_supported_standards](#icrc25_supported_standards)
```

**Review by frederikrothenberger — COMMENTED** [review]
Thanks a lot @dostro! I think this is a very good step forward. 🙂 
I found only minor errors, most comments are only suggestions to clarify things.

Given you are renaming the ICRC-28 file, please also adjust the link in the `README.md`.

File: topics/icrc_34_delegation.md (hunk 12 above)
Comment (dostro) [reply topics/icrc_34_delegation.md:38]: updated

File: topics/icrc_34_delegation.md (hunk 20 above)
Comment (dostro) [reply topics/icrc_34_delegation.md:152]: Added

File: topics/icrc_28_trusted_origins.md [hunk 22]
```diff
@@ -6,166 +6,77 @@
 <!-- TOC -->
 * [ICRC-28: Trusted Origins](#icrc-28-trusted-origins)
   * [Summary](#summary)
-  * [Identify](#identify)
-  * [Verify](#verify)
-    * [icrc28_get_trusted_origins](#icrc28_get_trusted_origins)
-  * [Supported standards](#supported-standards)
-    * [icrc28_supported_standards](#icrc28_supported_standards)
+  * [Method](#method)
```
[the remark this answers is not in the document]
Comment (sea-snake) [reply topics/icrc_28_trusted_origins.md:9]: It looks like you forgot it too in ICRC-21 😅

[the remark this answers is not in the document]
Comment (sea-snake) [reply topics/icrc_28_trusted_origins.md:9]: > Please also add a section referring to `icrc10_supported_standards` (i.e. ICRC-28 should be listed there).
It looks like you forgot it too in ICRC-21 😅

[the remark this answers is not in the document]
Comment (dostro) [reply topics/icrc_28_trusted_origins.md]: added

File: topics/icrc_28_trusted_origins.md [hunk 23]
```diff
@@ -16,25 +17,37 @@
...
 
+### icrc10_supported_standards
+
+An implementation of ICRC-28 MUST implement the method icrc10_supported_standards as put forth in 
+[ICRC-10](https://github.com/dfinity/ICRC/ICRCs/ICRC-10).
+
+The result of the call MUST always have at least the following entries:
+
+```
+record { name = "ICRC-28"; url = "https://github.com/dfinity/ICRC/ICRCs/ICRC-28"; }
```
Comment (frederikrothenberger) [inline topics/icrc_28_trusted_origins.md:47]: I'm really not a fan of the `url` entries here, since they do _not_ point anywhere sensible. Imho, at least the entry for ICRC-28 should be a correct URL, which would then be `https://github.com/dfinity/ICRC/tree/main/ICRCs/ICRC-28/ICRC-28.md`...

File: topics/icrc_34_delegation.md [hunk 24]
```diff
@@ -156,11 +157,37 @@ the [IC interface specification, authentication section](https://internetcompute
...
+
+```mermaid
+  sequenceDiagram
+    participant RP as Relying Party
+    participant S as Signer
+    participant C as Target Canister
+    RP ->> S: Request delegation
+    alt Relying party has not been granted <br>the `icrc34_delegation` permission scope<br>or the request does not comply with scope restrictions
+        S ->> RP: Error response: Permission not granted (3000)
+    else Requests includes targets
```
Comment (frederikrothenberger) [inline topics/icrc_34_delegation.md:174]:
```suggestion
    else Requests includes targets and signer supports account delegations
```

File: topics/icrc_28_trusted_origins.md [hunk 25]
```diff
@@ -45,11 +58,29 @@ with icrc28-conforming canisters can call those canisters without user approval.
 ### Account Delegation Use-Case
 
 [ICRC-34](./icrc_34_delegation.md) applies when a user returns a delegation to a relying party that 
-can be used to make authenticated calls on the user's behalf (i.e. without displaying wallet approval prompts). 
-To do this safely for the user's wallet address in a way that prevents malicious actors from having access to 
+can be used to make authenticated calls on the user's behalf (i.e. without displaying wallet approval prompts).
+Reminder that delegations come in two forms:
```
Comment (frederikrothenberger) [inline topics/icrc_28_trusted_origins.md:62]:
```suggestion
These delegations come in two forms:
```

**Review by frederikrothenberger — APPROVED** [review]
Thanks @dostro, great work!

I only found a few minor things.

File: topics/icrc_28_trusted_origins.md (hunk 23 above)
Comment (dostro) [reply topics/icrc_28_trusted_origins.md]: updated (and created a PR for icrc28 placeholder in ICRCs repo
